### PR TITLE
feat: adds some common argo-workflows workflow templates

### PR DESCRIPTION
**Pull request by andrrax**
This adds some common generic workflow templates, bumps argo-workflows to 2.5.6, as well as adjust the argo-role to allow for http template usage as well as memoization.

### CODE REVIEW

**Review by cardoe — COMMENTED** [review]
So I think this is good. There's some overlap with Milan so it would be good for your guys to figure out a common path forward.

**Comment by cardoe** [comment]
We need to figure out a good way to do docs that's not random readmes in folders. It should be plumbed to the top level docs. Need to reconcile against #78

**Comment by mfencik** [comment]
> So I think this is good. There's some overlap with Milan so it would be good for your guys to figure out a common path forward.

Where is the overlap?

**Review by skrobul — COMMENTED** [review]
Good job, I like the clever solution with storing credentials as Secrets for the pod lifetime. Caught few typos and had some minor questions/suggestions above.

File: argo-workflows/generic/docs/README.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+which can then be consumed by future WorkflowTemplates.
+
+## WorkflowTemplates
+
+| WorkflowTemplate | Description | Input | Output |
+|-------------------|-------------------------------------------------------|------------|------------|
+| get-device-nautobot | Return Device Information from Nautobot | hostname | device |
+| get-obm-creds | Get the credentials for the target Device | hostname | secret |
+| get-obm-ip | Get OBM IP address for target Device | hostname | ip |
+| nautobot-api | HTTP Template Workflow to query the Nautobot API | method,nautobot_url,uri | result |
```
[the remark this answers is not in the document]
Comment (andrrax) [reply argo-workflows/generic/docs/README.md:14]: Is this just a formatting/alignment thing? I'm not seeing any other grammatical issues here. I believe markdown uses the hyphenated line to set column widths, in which case both of these will render the same way. Maybe the suggested approach is stylistically preferred?

File: argo-workflows/generic/workflowtemplates/nautobot-api.yaml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                - name: url
+                  value: "{{inputs.parameters.nautobot_url}}/api{{inputs.parameters.uri}}"
+
+    - name: http
+      inputs:
+        parameters:
+          - name: method
+          - name: url
+      http:
+        timeoutSeconds: 30
```
[the remark this answers is not in the document]
Comment (andrrax) [reply argo-workflows/generic/workflowtemplates/nautobot-api.yaml:50]: I looked into this approach, and ran into a limitation of argo-workflows I believe. When trying to set timeoutSeconds to "{{inputs.parameters.timeout}}", it throws:

`FATA[2024-05-07T16:25:52.831Z] Failed to submit workflow: rpc error: code = InvalidArgument desc = json: cannot unmarshal string into Go struct field HTTP.spec.templates.http.timeoutSeconds of type int64 `

argo doesn't appear to handle casting to int (with int(), asInt()) properly there either. Looks like a known issue:

https://github.com/argoproj/argo-workflows/issues/1780

Maybe something for down the road.

[the remark this answers is not in the document]
Comment (skrobul) [reply argo-workflows/generic/workflowtemplates/nautobot-api.yaml:50]: Ouch, that's surprising - I would have expected it to accept strings and do the conversion itself. Guess we'll just increase it when needed. 30 seconds should cover everything for now, thanks!

File: argo-workflows/generic/docs/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (skrobul) [reply argo-workflows/generic/docs/README.md:14]: Yeah, just whitespace that will eventually render to same final product -  it just makes it easier to read and edit outside of the browser

File: .github/workflows/build-container-images.yaml [hunk 3]
```diff
@@ -7,18 +7,22 @@ on:
     paths:
       - 'argo-workflows/generic/containers/*'
       - 'argo-workflows/ironic-nautobot-sync/containers/*'
+      - 'argo-workflows/obm-utils/containers/*'
```
Comment (skrobul) [inline .github/workflows/build-container-images.yaml:10]: It's not related only to your PR/feature, but do you think we should also trigger this on any change under the `code/` folder too?

File: .github/workflows/build-container-images.yaml [hunk 4]
```diff
@@ -71,3 +75,25 @@ jobs:
...
               org.opencontainers.image.version=${{ env.VERSION_PYTHON_IRONIC }}
+
+      - name: Build and deploy Argo Utils image
+        uses: docker/build-push-action@v5
+        with:
+          context: argo-workflows/generic/
+          file: argo-workflows/generic/containers/Dockerfile.argo_utils
+          # push for all main branch commits
+          push: ${{ github.event_name != 'pull_request' }}
+          tags: ghcr.io/rackerlabs/understack/argo-utils-python3.11.8:latest,ghcr.io/rackerlabs/understack/argo-utils-python3.11.8:${{ env.VERSION_ARGO_UTILS }}
```
Comment (skrobul) [inline .github/workflows/build-container-images.yaml:86]: Please have a look at a [discussion around including python version in non-base containers](https://github.com/rackerlabs/understack/pull/78#discussion_r1591640481) and let us know what you think. If you agree, I'll create separate PR to adjust the `ironic-nautobot-sync` image name.

**Review by skrobul — APPROVED** [review]

**Pull request merged**